src/librustc_mir/transform/move_up_propagation.rs

```diff
@@ -0,0 +1,494 @@
+                        .or_insert(BitVector::new(num_statements))
+                        .insert(use_idx);
+                    continue;
+                }
+            }
+            panic!("We should have already checked for this");
+        }
+
+        {
+            let bbs = mir.basic_blocks_mut();
+            for (bb, repls) in old_2_new {
+                let new_stmts: Vec<_> = bbs[bb].statements
+                                                .iter()
+                                                .enumerate()
+                                                .map(|(stmt_idx, orig_stmt)| {
+                    if let Some(repl) = repls.get(&stmt_idx) {
+                        let repl_stmt = Statement { kind: repl.clone(),
+                                                    source_info: orig_stmt.source_info,
+                        };
+                        debug!("replacing {:?} with {:?}", orig_stmt, repl_stmt);
+                        repl_stmt
+                    } else {
+                        //debug!("repl idx: {:?} didnt match {:?}", idx, stmt_idx);
+                        orig_stmt.clone()
+                    }
+                }).collect();
+                bbs[bb] = BasicBlockData {
+                    statements: new_stmts,
+                    terminator: bbs[bb].terminator.clone(),
+                    is_cleanup: bbs[bb].is_cleanup,
+                };
+            }
+
+            for (bb, dead_idxs) in dead {
+                let mut idx_cnt = 0;
+                bbs[bb].statements.retain(|_| {
+                    let dead = dead_idxs.contains(idx_cnt);
+                    idx_cnt += 1;
+                    !dead
+                });
+            }
+        }
+
+    }
+}
+
+fn get_next_locs(curr: UseDefLocation, mir: &Mir) -> Vec<UseDefLocation> {
+    match curr.inner_location {
+        InnerLocation::Terminator => {
+            mir.basic_blocks()[curr.basic_block].terminator().successors().iter().map(|&s| {
+                UseDefLocation {
+                    basic_block: s,
+                    inner_location: InnerLocation::StatementIndex(0),
+                }
+            }).collect()
+        }
+        InnerLocation::StatementIndex(idx) => {
+            if idx + 1 < mir.basic_blocks()[curr.basic_block].statements.len() {
+                vec![UseDefLocation{
+                    basic_block: curr.basic_block,
+                    inner_location: InnerLocation::StatementIndex(idx + 1),
+                }]
+            } else {
+                let next = UseDefLocation{ basic_block: curr.basic_block,
+                                           inner_location: InnerLocation::Terminator,
+                };
+                get_next_locs(next, mir)
+            }
+        }
+    }
+}
+
+fn paths_contain_call(start: UseDefLocation,
+                      target: UseDefLocation,
+                      mir: &Mir,
+                      visited: &mut HashMap<UseDefLocation, bool>)
```

Comment:
BitVector please. Also, you are exactly the `O(n^2)` here that we were trying to avoid with postdominators.